chain/store/store.go

```diff
@@ -1052,7 +1052,15 @@ func (cs *ChainStore) expandTipset(ctx context.Context, b *types.BlockHeader) (*
 		}
 
 		if cid, found := inclMiners[h.Miner]; found {
-			log.Warnf("Have multiple blocks from miner %s at height %d in our tipset cache %s-%s", h.Miner, h.Height, h.Cid(), cid)
+			log.Warnf("Have multiple blocks from miner %s at height %d in our tipset cache %s-%s, purging them all", h.Miner, h.Height, h.Cid(), cid)
+			newAll := make([]*types.BlockHeader, 0, len(all))
+			for _, blk := range all {
+				if blk.Miner != h.Miner {
+					newAll = append(newAll, blk)
```

Comment:
nit: I'd just collect all these blocks into a miner -> block map, then put them into an array at the very end.